src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -1033,6 +1077,58 @@ private XmlSerializationFixupCallback CreateWriteFixupMethod(Member[] members)
             };
         }
 
+        private void WriteAddCollectionFixup(object o, Member member, object memberValue)
+        {
+            TypeDesc typeDesc = member.Mapping.TypeDesc;
+            bool readOnly = member.Mapping.ReadOnly;
+            object memberSource = GetMemberValue(o, member.Mapping.MemberInfo);
+            if (memberSource == null)
+            {
+                if (readOnly)
+                {
+                    throw CreateReadOnlyCollectionException(typeDesc.CSharpName);
+                }
+
+                memberSource = ReflectionCreateObject(typeDesc.Type);
+                SetMemberValue(o, memberSource, member.Mapping.MemberInfo);
+            }
+
+            var collectionFixup = new CollectionFixup(
+                memberSource,
+                new XmlSerializationCollectionFixupCallback(GetCreateCollectionOfObjectsCallback(typeDesc.Type)),
+                memberValue);
+
+            AddFixup(collectionFixup);
+        }
+
+        private XmlSerializationCollectionFixupCallback GetCreateCollectionOfObjectsCallback(Type collectionType)
+        {
+            return (collection, collectionItems) =>
+            {
+                if (collectionItems == null)
+                    return;
+
+                if (collection == null)
+                    return;
+
+                var listOfItems = new List<object>();
+                var enumerableItems = collectionItems as IEnumerable;
```

Comment:
What's the performance like for this? Iterating over an IEnumerable has some overhead. If the collection is an IList, then casting to that can produce faster code. It would be worth benchmarking the difference in performance when collectionItems is a list.